tests/test_stateful_browser.py

```diff
@@ -351,5 +353,65 @@ def test_select_form_nr():
             browser.select_form(nr=3)
 
 
+def test_download_link(httpbin):
+    """Test downloading the contents of a link to file."""
+    browser = mechanicalsoup.StatefulBrowser()
+    browser.open(httpbin.url)
+    tmpfile = tempfile.NamedTemporaryFile()
+    current_url = browser.get_url()
+    current_page = browser.get_current_page()
+    browser.download_link(tmpfile.name, link='image/png')
+
+    # Check that the browser state has not changed
+    assert browser.get_url() == current_url
+    assert browser.get_current_page() == current_page
+
+    # Check that the file was downloaded
+    assert os.path.isfile(tmpfile.name)
+
+
+def test_download_link_404(httpbin):
+    """Test downloading the contents of a broken link."""
+    browser = mechanicalsoup.StatefulBrowser(raise_on_404=True)
+    browser.open_fake_page('<a href="/no-such-page-404">Link</a>',
+                           url=httpbin.url)
+    tmpdir = tempfile.mkdtemp()
+    tmpfile = tmpdir + '/nosuchfile.txt'
+    current_url = browser.get_url()
+    current_page = browser.get_current_page()
+    with pytest.raises(mechanicalsoup.LinkNotFoundError):
+        browser.download_link(tmpfile, link_text='Link')
+
+    # Check that the browser state has not changed
+    assert browser.get_url() == current_url
+    assert browser.get_current_page() == current_page
+
+    # Check that the file was not downloaded
+    assert not os.path.exists(tmpfile)
+
+
+def test_download_link_referer(httpbin):
+    """Test downloading the contents of a link to file."""
+    browser = mechanicalsoup.StatefulBrowser()
+    ref = httpbin.url + "/my-referer"
```

Comment:
Very minor, but the `httpbin` object implements an `operator+` for strings, you can just use `httpbin + "my-referer"` to achieve the same thing. Feel free to ignore. :)

Reply:
Done.